### PR TITLE
storage service update handler Implements

### PR DESCRIPTION
**Pull request by soulseen**
Signed-off-by: Xiaoyang Zhu <zhuxiaoyang1996@gmail.com>

### CODE REVIEW

File: controllers/storage/minio.go [hunk 1]
```diff
@@ -50,7 +50,19 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
 	var minioCR minio.MinIOInstance
 
 	if m.HarborCluster.Spec.Storage.Kind != inClusterStorage {
-		return m.ProvisionExternalStorage()
+		var exSecret corev1.Secret
+		err := m.KubeClient.Get(m.getExternalSecretNamespacedName(), &exSecret)
```
Comment (soulseen) [inline controllers/storage/minio.go:54]: Couple  things I think we should aware of:
1. switch external to inCluster 
1. switch inCluster to external

Comment (soulseen) [reply controllers/storage/minio.go]: Implements it in the next version.

File: controllers/storage/provision.go [hunk 2]
```diff
@@ -59,74 +45,75 @@ func (m *MinIOReconciler) generateInClusterSecret(minioInstamnce *minio.MinIOIns
 		},
 		Type: corev1.SecretTypeOpaque,
 		Data: map[string][]byte{
-			// TODO how about using random passwords ??
+			// TODO how about using random password ??
```
Comment (steven-zou) [inline controllers/storage/provision.go:48]: +1 for using random pwd

File: controllers/storage/minio.go [hunk 3]
```diff
@@ -96,7 +135,15 @@ func createDefaultBucket() error {
 }
 
 func (m *MinIOReconciler) checkMinIOUpdate() bool {
-	panic("implement me")
+	if !cmp.Equal(m.DesiredMinIOCR.DeepCopy().Spec, m.CurrentMinIOCR.DeepCopy().Spec) {
```
Comment (wangcanfengxs) [inline controllers/storage/minio.go:138]: If `CurrentMinioCR` has some default value, then this `if` is alway true.

File: controllers/cache/utils.go [hunk 4]
```diff
@@ -36,6 +36,7 @@ func generateName(typeName, metaName string) string {
 	return fmt.Sprintf("%s-%s", typeName, metaName)
 }
 
+// Deprecated
```
Comment (steven-zou) [inline controllers/cache/utils.go:39]: So @Ghostbaby should replace this function with the new common util func?

File: controllers/common/utils.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	if strings.Contains(randType, "a") {
+		b.WriteString(lower)
+	}
+	if strings.Contains(randType, "A") {
+		b.WriteString(upper)
+	}
+	var str = b.String()
+	var strLen = len(str)
+	if strLen == 0 {
+		result = ""
```
Comment (steven-zou) [inline controllers/common/utils.go:28]: For a secret generation, it seems empty str is not accepted, right?

Comment (steven-zou) [reply controllers/common/utils.go:28]: Maybe an error should be returned.

File: controllers/storage/minio.go [hunk 6]
```diff
@@ -83,20 +119,28 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
...
-		return m.ProvisionInClusterSecretAsOss(&minioCR)
+		return m.ProvisionInClusterSecretAsS3(&minioCR)
 	}
 
-	return nil, nil
+	return minioUnknownStatus(), nil
 }
 
 func createDefaultBucket() error {
 	panic("implement me")
```
Comment (steven-zou) [inline controllers/storage/minio.go:131]: Why not implemented it?

File: controllers/cache/utils.go (hunk 4 above)
Comment (soulseen) [reply controllers/cache/utils.go:39]: yes, it moved to common util.

File: controllers/storage/minio.go (hunk 6 above)
Comment (soulseen) [reply controllers/storage/minio.go:131]: sorry,  I forgot.  I will create another PR to implement it.

File: controllers/cache/utils.go (hunk 4 above)
Comment (Ghostbaby) [reply controllers/cache/utils.go:39]: ok, I will remove this function.

**Pull request merged**